File: application/credit-card-enrollment-backend-in-java/src/main/java/cloud/ambar/common/serializedevent/Deserializer.java
package cloud.ambar.common.serializedevent;

import cloud.ambar.common.event.Event;
import cloud.ambar.creditcard.enrollment.event.EnrollmentDeclined;
import cloud.ambar.creditcard.enrollment.event.EnrollmentSubmittedForReview;
import cloud.ambar.creditcard.enrollment.event.EnrollmentRequested;
import cloud.ambar.creditcard.product.event.ProductActivated;
import cloud.ambar.creditcard.product.event.ProductDeactivated;
import cloud.ambar.creditcard.product.event.ProductDefined;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.stereotype.Service;

import java.time.Instant;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;

@Service
public class Deserializer {
    public Event deserialize(SerializedEvent serializedEvent) {
        return switch (serializedEvent.getEventName()) {
            case "CreditCard_Enrollment_EnrollmentRequested" -> EnrollmentRequested.builder()
                    .eventId(serializedEvent.getEventId())
                    .aggregateId(serializedEvent.getAggregateId())
                    .aggregateVersion(serializedEvent.getAggregateVersion())
                    .correlationId(serializedEvent.getCorrelationId())
                    .causationId(serializedEvent.getCausationId())
                    .recordedOn(toInstant(serializedEvent.getRecordedOn()))
                    .annualIncomeInCents(payloadInt(serializedEvent.getJsonPayload(), "annualIncomeInCents"))
                    .productId(payloadString(serializedEvent.getJsonPayload(), "productId"))
                    .userId(payloadString(serializedEvent.getJsonPayload(), "userId"))
                    .build();
            case "CreditCard_Enrollment_EnrollmentSubmittedForReview" -> EnrollmentSubmittedForReview.builder()
                    .eventId(serializedEvent.getEventId())
                    .aggregateId(serializedEvent.getAggregateId())
                    .aggregateVersion(serializedEvent.getAggregateVersion())
                    .correlationId(serializedEvent.getCorrelationId())
                    .causationId(serializedEvent.getCausationId())
                    .recordedOn(toInstant(serializedEvent.getRecordedOn()))
                    .build();
            case "CreditCard_Enrollment_EnrollmentDeclined" -> EnrollmentDeclined.builder()
                    .eventId(serializedEvent.getEventId())
                    .aggregateId(serializedEvent.getAggregateId())
                    .aggregateVersion(serializedEvent.getAggregateVersion())
                    .correlationId(serializedEvent.getCorrelationId())
                    .causationId(serializedEvent.getCausationId())
                    .recordedOn(toInstant(serializedEvent.getRecordedOn()))
                    .reasonCode(payloadString(serializedEvent.getJsonPayload(), "reasonCode"))
                    .reasonDescription(payloadString(serializedEvent.getJsonPayload(), "reasonDescription"))
                    .build();
            case "CreditCard_Product_ProductActivated" -> ProductActivated.builder()
                    .eventId(serializedEvent.getEventId())
                    .aggregateId(serializedEvent.getAggregateId())
                    .aggregateVersion(serializedEvent.getAggregateVersion())
                    .correlationId(serializedEvent.getCorrelationId())
                    .causationId(serializedEvent.getCausationId())
                    .recordedOn(toInstant(serializedEvent.getRecordedOn()))
                    .build();
            case "CreditCard_Product_ProductDeactivated" -> ProductDeactivated.builder()
                    .eventId(serializedEvent.getEventId())
                    .aggregateId(serializedEvent.getAggregateId())
                    .aggregateVersion(serializedEvent.getAggregateVersion())
                    .correlationId(serializedEvent.getCorrelationId())
                    .causationId(serializedEvent.getCausationId())
                    .recordedOn(toInstant(serializedEvent.getRecordedOn()))
                    .build();
            case "CreditCard_Product_ProductDefined" -> ProductDefined.builder()
                    .eventId(serializedEvent.getEventId())
                    .aggregateId(serializedEvent.getAggregateId())
                    .aggregateVersion(serializedEvent.getAggregateVersion())
                    .correlationId(serializedEvent.getCorrelationId())
                    .causationId(serializedEvent.getCausationId())
                    .recordedOn(toInstant(serializedEvent.getRecordedOn()))
                    .name(payloadString(serializedEvent.getJsonPayload(), "name"))
                    .interestInBasisPoints(payloadInt(serializedEvent.getJsonPayload(), "interestInBasisPoints"))
                    .annualFeeInCents(payloadInt(serializedEvent.getJsonPayload(), "annualFeeInCents"))
                    .paymentCycle(payloadString(serializedEvent.getJsonPayload(), "paymentCycle"))
                    .creditLimitInCents(payloadInt(serializedEvent.getJsonPayload(), "creditLimitInCents"))
                    .maxBalanceTransferAllowedInCents(payloadInt(serializedEvent.getJsonPayload(), "maxBalanceTransferAllowedInCents"))
                    .reward(payloadString(serializedEvent.getJsonPayload(), "reward"))
                    .cardBackgroundHex(payloadString(serializedEvent.getJsonPayload(), "cardBackgroundHex"))
                    .build();
            default -> throw new RuntimeException("Unknown event type: " + serializedEvent.getEventName());
        };
    }


    private Instant toInstant(String recordedOn) {
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss.SSSSSS z");
        return ZonedDateTime.parse(recordedOn, formatter).toInstant();
    }

    private String payloadString(String jsonString, String fieldName) {
        final ObjectMapper objectMapper = new ObjectMapper();

        try {
            return objectMapper.readTree(jsonString).get(fieldName).asText();
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    private int payloadInt(String jsonString, String fieldName) {
        final ObjectMapper objectMapper = new ObjectMapper();

        try {
            return objectMapper.readTree(jsonString).get(fieldName).asInt();
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }
}

File: application/backend-credit-card-enrollment/backend-c#/CreditCardEnrollment/Common/Events/Deserializer.cs
using System.Text.Json;
using System.Text.Json.Nodes;
using CreditCardEnrollment.Domain.Enrollment.Events;
using CreditCardEnrollment.Domain.Product.Events;

namespace CreditCardEnrollment.Common.Events;

public interface IDeserializer
{
    Event Deserialize(SerializedEvent serializedEvent);
}

public class Deserializer : IDeserializer
{
    public Event Deserialize(SerializedEvent serializedEvent)
    {
        return serializedEvent.EventName switch
        {
            "CreditCard_Enrollment_EnrollmentRequested" => new EnrollmentRequested
            {
                EventId = serializedEvent.EventId,
                AggregateId = serializedEvent.AggregateId,
                AggregateVersion = serializedEvent.AggregateVersion,
                CorrelationId = serializedEvent.CorrelationId,
                CausationId = serializedEvent.CausationId,
                RecordedOn = ToDateTime(serializedEvent.RecordedOn),
                AnnualIncomeInCents = PayloadInt(serializedEvent.JsonPayload, "annualIncomeInCents"),
                ProductId = PayloadString(serializedEvent.JsonPayload, "productId"),
                UserId = PayloadString(serializedEvent.JsonPayload, "userId")
            },
            "CreditCard_Enrollment_EnrollmentAccepted" => new EnrollmentAccepted
            {
                EventId = serializedEvent.EventId,
                AggregateId = serializedEvent.AggregateId,
                AggregateVersion = serializedEvent.AggregateVersion,
                CorrelationId = serializedEvent.CorrelationId,
                CausationId = serializedEvent.CausationId,
                RecordedOn = ToDateTime(serializedEvent.RecordedOn),
                ReasonDescription = PayloadString(serializedEvent.JsonPayload, "reasonDescription"),
                UserId = PayloadString(serializedEvent.JsonPayload, "userId"),
                ProductId = PayloadString(serializedEvent.JsonPayload, "productId")
            },
            "CreditCard_Enrollment_EnrollmentDeclined" => new EnrollmentDeclined
            {
                EventId = serializedEvent.EventId,
                AggregateId = serializedEvent.AggregateId,
                AggregateVersion = serializedEvent.AggregateVersion,
                CorrelationId = serializedEvent.CorrelationId,
                CausationId = serializedEvent.CausationId,
                RecordedOn = ToDateTime(serializedEvent.RecordedOn),
                Reason = PayloadString(serializedEvent.JsonPayload, "reasonDescription"),
                UserId = PayloadString(serializedEvent.JsonPayload, "userId"),
                ProductId = PayloadString(serializedEvent.JsonPayload, "productId")
            },
            "CreditCard_Product_ProductActivated" => new ProductActivated
            {
                EventId = serializedEvent.EventId,
                AggregateId = serializedEvent.AggregateId,
                AggregateVersion = serializedEvent.AggregateVersion,
                CorrelationId = serializedEvent.CorrelationId,
                CausationId = serializedEvent.CausationId,
                RecordedOn = ToDateTime(serializedEvent.RecordedOn)
            },
            "CreditCard_Product_ProductDeactivated" => new ProductDeactivated
            {
                EventId = serializedEvent.EventId,
                AggregateId = serializedEvent.AggregateId,
                AggregateVersion = serializedEvent.AggregateVersion,
                CorrelationId = serializedEvent.CorrelationId,
                CausationId = serializedEvent.CausationId,
                RecordedOn = ToDateTime(serializedEvent.RecordedOn)
            },
            "CreditCard_Product_ProductDefined" => new ProductDefined
            {
                EventId = serializedEvent.EventId,
                AggregateId = serializedEvent.AggregateId,
                AggregateVersion = serializedEvent.AggregateVersion,
                CorrelationId = serializedEvent.CorrelationId,
                CausationId = serializedEvent.CausationId,
                RecordedOn = ToDateTime(serializedEvent.RecordedOn),
                Name = PayloadString(serializedEvent.JsonPayload, "name"),
                InterestInBasisPoints = PayloadInt(serializedEvent.JsonPayload, "interestInBasisPoints"),
                AnnualFeeInCents = PayloadInt(serializedEvent.JsonPayload, "annualFeeInCents"),
                PaymentCycle = PayloadString(serializedEvent.JsonPayload, "paymentCycle"),
                CreditLimitInCents = PayloadInt(serializedEvent.JsonPayload, "creditLimitInCents"),
                MaxBalanceTransferAllowedInCents = PayloadInt(serializedEvent.JsonPayload, "maxBalanceTransferAllowedInCents"),
                Reward = PayloadString(serializedEvent.JsonPayload, "reward"),
                CardBackgroundHex = PayloadString(serializedEvent.JsonPayload, "cardBackgroundHex")
            },
            _ => throw new ArgumentException($"Unknown event type: {serializedEvent.EventName}")
        };
    }

    private static DateTime ToDateTime(string? recordedOn)
    {
        if (string.IsNullOrEmpty(recordedOn))
            return DateTime.UtcNow;

        // Parse the date format "yyyy-MM-dd HH:mm:ss.SSSSSS z"
        if (DateTime.TryParse(recordedOn, out DateTime result))
            return result;

        throw new ArgumentException($"Invalid date format: {recordedOn}");
    }

    private static string PayloadString(string jsonString, string fieldName)
    {
        try
        {
            var jsonNode = JsonNode.Parse(jsonString);
            return jsonNode?[fieldName]?.GetValue<string>() ?? string.Empty;
        }
        catch (Exception ex)
        {
            throw new ArgumentException($"Error parsing JSON field {fieldName}", ex);
        }
    }

    private static int PayloadInt(string jsonString, string fieldName)
    {
        try
        {
            var jsonNode = JsonNode.Parse(jsonString);
            return jsonNode?[fieldName]?.GetValue<int>() ?? 0;
        }
        catch (Exception ex)
        {
            throw new ArgumentException($"Error parsing JSON field {fieldName}", ex);
        }
    }
}
